pkg/flowkit/gateway/emulator.go

```diff
@@ -19,31 +19,47 @@
 package gateway
 
 import (
-	"fmt"
-	"time"
-
-	"github.com/onflow/flow-cli/pkg/flowkit"
-	"github.com/onflow/flow-cli/pkg/flowkit/config"
-
-	emulator "github.com/onflow/flow-emulator"
-	flowGo "github.com/onflow/flow-go/model/flow"
+	"context"
 
 	"github.com/onflow/cadence"
+	emulator "github.com/onflow/flow-emulator"
+	"github.com/onflow/flow-emulator/server/backend"
 	"github.com/onflow/flow-go-sdk"
 	"github.com/onflow/flow-go-sdk/client"
 	"github.com/onflow/flow-go-sdk/client/convert"
+	flowGo "github.com/onflow/flow-go/model/flow"
+	"github.com/sirupsen/logrus"
+
+	"github.com/onflow/flow-cli/pkg/flowkit"
+	"github.com/onflow/flow-cli/pkg/flowkit/config"
 )
 
 type EmulatorGateway struct {
 	emulator *emulator.Blockchain
+	backend  *backend.Backend
+	ctx      context.Context
 }
 
 func NewEmulatorGateway(serviceAccount *flowkit.Account) *EmulatorGateway {
+	return NewEmulatorGatewayWithLogger(logrus.New(), serviceAccount)
+}
+
+func NewEmulatorGatewayWithLogger(logger *logrus.Logger, serviceAccount *flowkit.Account) *EmulatorGateway {
```

Comment:
Should we maybe have functional options as the last argument and just have a factory with options, so for example:
```go
NewEmulatorWithOpts(serviceAccount *flowkit.Account, ...func(*EmulatorGateway)) { ... }

NewEmulatorWithOpts(service, WithLogger(lorgus.new())
```
Just so if we in the future need to have more options.

Reply:
Also we can then have `WithContext` and no need for bellow method, although I don't think this is a good pattern in Go, context should be passed as first argument, and I know I did that mistake myself when starting (since I didn't know enough Go), I would like to refactor this before making flowkit more official, but this is a task for another PR

Reply:
I see this `NewEmulator(service).WithLogger(logrus.new())` pattern used commonly but not sure which one is better tbh, but I can refactor to whichever we choose. 

I agree with context comment, actually I started with `WithContext`, then refactored it to `SetContext`, but to confess I am not too familiar with context there. I mean does contexts are has to be per call vs per lifecycle of the gateway.

Reply:
> I see this NewEmulator(service).WithLogger(logrus.new()) pattern used commonly but not sure which one is better tbh, but I can refactor to whichever we choose.

We can make it `.WithLogger()` as well as a method on the emulator, the only difference I would say is that if the option is needed during the initialization then I think it's better to have an option being passed to the factory method. I would say that for the logger it makes sense to be passed during the init but I can also see the case where you would just set it for the function you care about and then in a way disable it, but then also we could argue that's something that should be handled on the logger instance itself and verbosity settings etc not by setting it null. 

> I agree with the context comment, actually, I started with WithContext, then refactored it to SetContext, but to confess I am not too familiar with context there. I mean does contexts are has to be per call vs per lifecycle of the gateway.

In go we should pass context as the first argument, that's idiomatic go and the only reason this isn't done like that here is that when I was defining the interface for the gateway I didn't know better. I would want at some later point (maybe flowkit update we discuss) to make it correct.